### PR TITLE
Refractor build_arg_string to also deal with infile and outfile

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

Address the comment in https://github.com/GenericMappingTools/pygmt/issues/1832#issuecomment-1072974748.

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

File: pygmt/helpers/utils.py [hunk 1]
```diff
@@ -213,6 +217,10 @@ def build_arg_string(kwargs):
                 # remove any spaces in PROJ4 string
                 _value = str(kwargs[key]).replace(" ", "")
             gmt_args.append(rf"-{key}{_value}")
+    if infile:
+        gmt_args.append(str(infile))
+    if outfile:
+        gmt_args.append("->" + str(outfile))
     return " ".join(sorted(gmt_args))
```
Comment (seisman) [inline pygmt/helpers/utils.py:224]: I don't like how the arguments are sorted. For example, currently, the returned argument string may look like:
```
->outfile -A0 -Bafg -Ctopo -Vd infile
```
but I expect to see a more readable argument string like:
```
infile -A0 -Bafg -Ctopo -Vd ->outfile
```
which is more friendly when debugging.

I can rewrite these lines to 
```
gmt_args = sorted(gmt_args)
if infile:
    gmt_args = [str(infile)] + gmt_args
if outfile:
    gmt_args = gmt_args.append("->" + str(outfile))
return " ".join(gmt_args)
```
but I'm wondering if someone can give a better and cleaner solution.

File: pygmt/helpers/utils.py [hunk 2]
```diff
@@ -119,7 +119,7 @@ def dummy_context(arg):
     yield arg
 
 
-def build_arg_string(kwargs):
+def build_arg_string(kwargs, infile=None, outfile=None):
```
Comment (weiji14) [inline pygmt/helpers/utils.py:122]: In Python, kwargs is usually placed at the last position, but here it is placed as first. Maybe we should change the ordering of the parameters here? See also @meghanrjones's comment at https://github.com/GenericMappingTools/pygmt/pull/731#discussion_r825477816 and https://peps.python.org/pep-3102 about keyword-only arguments.

Comment (seisman) [reply pygmt/helpers/utils.py:122]: Actually, it's just a dict which has a name `kwargs`. Maybe `kwargs` is not a good name here.

Comment (weiji14) [reply pygmt/helpers/utils.py:122]: How about `kwdict`?

File: pygmt/helpers/utils.py (hunk 1 above)
Comment (weiji14) [reply pygmt/helpers/utils.py:224]: Something like this? Based on https://stackoverflow.com/questions/3752240/join-string-and-none-string-using-optional-delimiter/3752317#3752317
```suggestion
    return " ".join(filter(None, [infile, sorted(gmt_args), outfile]))
```

File: pygmt/src/blockm.py [hunk 3]
```diff
@@ -51,8 +51,10 @@ def _blockm(block_method, data, x, y, z, outfile, **kwargs):
             with table_context as infile:
                 if outfile is None:
                     outfile = tmpfile.name
-                arg_str = " ".join([infile, build_arg_string(kwargs), "->" + outfile])
-                lib.call_module(module=block_method, args=arg_str)
+                lib.call_module(
+                    module=block_method,
+                    args=build_arg_string(kwargs, infile=infile, outfile=outfile),
```
Comment (weiji14) [inline pygmt/src/blockm.py:56]: Still having mixed feeling about having kwargs as the first argument here, but I guess it works :upside_down_face:

**Review by weiji14 — APPROVED** [review]

**Comment by seisman** [comment]
This PR will conflict with PR #1857 or #1862, due to changes in grd2cpt and makecpt. We need to decide which PR should be merged first.

**Comment by weiji14** [comment]
> This PR will conflict with PR #1857 or #1862, due to changes in grd2cpt and makecpt. We need to decide which PR should be merged first.

Merge this one (#1837) first, so that you can proceed with #1831/#1832. The conflicts in #1857 and #1862 can be resolved later.

**Pull request merged**